lmp/infer/_top_p.py

```diff
@@ -0,0 +1,294 @@
+
+        # Convert to tensor with `dtype == torch.int64`.
+        # Tensor shape: `(1, S')`.
+        # Tensor dtype: `torch.int64`.
+        batch_prev_tkids = torch.LongTensor(batch_prev_tkids)
+
+        # Remove `[eos]` token id since model is not trained to predict tokens
+        # after seeing `[eos]`.
+        # Tensor shape: `(1, S'-1)` or `(1, S)`.
+        # Tensor dtype: `torch.int64`.
+        batch_prev_tkids = batch_prev_tkids[..., :-1]
+
+        # Satisty maximum sequence length constraint.
+        # If sequence length is longer than constraint, then truncate tensor
+        # to have shape `(1, self.max_seq_len)`.
+        # Otherwise tensor shape remain the same.
+        batch_prev_tkids = batch_prev_tkids[..., :self.max_seq_len]
+
+        # Get model running device.
+        device = next(model.parameters()).device
+
+        # Move tensors to model running device.
+        batch_prev_tkids = batch_prev_tkids.to(device)
+
+        # Calculate how many token can be generate at most.
+        # `out_seq_len` satisfy `0 <= out_seq_len <= self.max_seq_len`.
+        out_seq_len = self.max_seq_len - batch_prev_tkids.size(1)
+
+        # Generate tokens.
+        for _ in range(out_seq_len):
+            # Get probability distribution with current token ids.
+            # Input tensor : Current token ids.
+            # Input shape  : `(1, S)`.
+            # Input dtype  : `torch.int64`.
+            # Output tensor: Next token ids probability distribution.
+            # Output shape : `(1, S, V)`.
+            # Output dtype : `torch.float32`.
+            batch_next_tkids_probs = model.pred(
+                batch_prev_tkids=batch_prev_tkids
+            )
+
+            # Get the last token id probability distribution.
+            # Only need the last token since we already know every previous
+            # token ids.
+            # Input tensor : Next token ids probability distribution.
+            # Input shape  : `(1, S, V)`.
+            # Input dtype  : `torch.float32`.
+            # Output tensor: The last next token id probability distribution.
+            # Output shape : `(1, V)`.
+            # Output dtype : `torch.float32`.
+            batch_next_tkid_probs = batch_next_tkids_probs[:, -1]
+
+            # Sort the probability distribution in descending order.
+            # `batch_topp_tkid_probs` shape  : `(1, V)`.
+            # `batch_topp_tkid_probs` dtype  : `torch.float32`.
+            # `batch_topp_tkid` tensor       : The top K next token id.
+            # `batch_topp_tkid` shape        : `(1, V)`.
+            # `batch_topp_tkid` dtype        : `torch.int64`.
+            (
+                batch_topp_tkid_probs,
+                batch_topp_tkid,
+            ) = batch_next_tkid_probs.sort(
+                dim=-1,
+                descending=True
+            )
+
+            # Use the token set that cover P probability mass as possible next
```

Comment:
Change to "Calculate cumulative distribution and retrieve indices with cumulative probability higher than ``P``."